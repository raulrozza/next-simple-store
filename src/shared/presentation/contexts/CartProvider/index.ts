export { default as useAddToCart } from './hooks/useAddToCart';
