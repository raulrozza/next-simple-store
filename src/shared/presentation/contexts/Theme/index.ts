export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
