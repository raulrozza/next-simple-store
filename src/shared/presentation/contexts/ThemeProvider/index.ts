export { default as useTheme } from './hooks/useTheme';
export { default as ThemeProvider } from './implementations/StyledComponentTheme';
