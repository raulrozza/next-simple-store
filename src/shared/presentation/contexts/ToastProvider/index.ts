export { useToastProvider } from './context';
export { default as ToastProvider } from './implementations/ReactToastifyProvider';
