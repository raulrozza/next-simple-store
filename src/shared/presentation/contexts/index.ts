export * from './CartProvider';
export * from './ThemeProvider';
