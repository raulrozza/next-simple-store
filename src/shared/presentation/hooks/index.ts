import { trpc } from '@/shared/infra/http/trpc';

const { useQuery, useMutation } = trpc;

export { useQuery, useMutation };
