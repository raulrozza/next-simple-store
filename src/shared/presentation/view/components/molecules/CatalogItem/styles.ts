import styled, { css } from 'styled-components';

export const Item = styled.li`
    ${({ theme }) => css`
        display: grid;
        grid-template-rows: auto ${theme.layout.size(3.5)} ${theme.layout.size(
                2,
            )} ${theme.layout.size(2)} ${theme.layout.size(2)};

        background-color: ${theme.palette.secondary['300']};

        overflow: hidden;
        height: ${theme.layout.size(23)};
        color: ${theme.palette.gray['0']};
        padding-bottom: ${theme.layout.spacing(1)};

        strong {
            font-family: ${theme.typography.family.title};
            font-size: ${theme.typography.sizes.lg};
            margin: ${theme.layout.spacing(1)};
        }

        .price {
            margin: ${theme.layout.spacing(0, 1)};
        }

        button {
            justify-self: center;
        }
    `}
`;

export const QuantityRow = styled.div`
    ${({ theme }) => css`
        display: flex;
        align-items: center;
        gap: ${theme.layout.spacing(1)};
        margin: ${theme.layout.spacing(1)};

        label {
            font-size: ${theme.typography.sizes.sm};
        }

        input {
            display: flex;
            width: 100%;
            outline: none;
            font-size: ${theme.typography.sizes.sm};
        }
    `}
`;
