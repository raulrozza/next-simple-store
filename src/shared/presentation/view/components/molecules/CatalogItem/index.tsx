import { FC, useCallback, useRef } from 'react';

import Image from 'next/image';

import IMG_PLACEHOLDER from '@/assets/img/product.jpg';
import { IProduct } from '@/shared/domain/entities/Product';
import { Button } from '@/shared/presentation/view/components/atoms';

import { Item, QuantityRow } from './styles';

const isQuantityValid = (quantity: number) => {
    return !isNaN(quantity) && quantity > 0;
};

const DEFAULT_QUANTITY = '1';

interface CatalogItemProps {
    product: IProduct & { imgAlt: string };
    onAddToCart: (params: { product: IProduct; quantity: number }) => void;
}

const CatalogItem: FC<CatalogItemProps> = ({ product, onAddToCart }) => {
    const inputRef = useRef<HTMLInputElement>(null);

    const handleAddToCart = useCallback(() => {
        if (!inputRef.current) return;

        const quantity = Number(inputRef.current.value.trim());

        if (!isQuantityValid(quantity)) return;

        onAddToCart({ product, quantity });
        inputRef.current.value = DEFAULT_QUANTITY;
    }, [onAddToCart, product]);

    return (
        <Item key={product.id}>
            <Image
                src={IMG_PLACEHOLDER}
                alt={product.imgAlt}
                layout="responsive"
            />

            <strong>{product.name}</strong>

            <p className="price">${product.price.toFixed(2)}</p>

            <QuantityRow>
                <label htmlFor={`quantity-${product.id}`}>Quantity</label>
                <input
                    id={`quantity-${product.id}`}
                    type="number"
                    ref={inputRef}
                    defaultValue={DEFAULT_QUANTITY}
                />
            </QuantityRow>

            <Button variant="secondary" icon="cart" onClick={handleAddToCart}>
                Add to cart
            </Button>
        </Item>
    );
};

export default CatalogItem;
