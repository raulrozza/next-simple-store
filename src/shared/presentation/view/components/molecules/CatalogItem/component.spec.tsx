import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ThemeProvider } from '@/shared/presentation/contexts';

import CatalogItem from './index';

describe('CatalogItem', () => {
    it('should render correctly', () => {
        const view = render(
            <CatalogItem
                product={{
                    id: 'id',
                    name: 'name',
                    description: 'description',
                    imgAlt: 'imgAlt',
                    price: 10,
                    slug: 'slug',
                }}
                onAddToCart={jest.fn()}
            />,
            {
                wrapper: ThemeProvider,
            },
        );

        expect(screen.getByText('name')).toBeInTheDocument();
        expect(screen.getByText('$10.00')).toBeInTheDocument();
        expect(screen.getByLabelText('Quantity')).toBeInTheDocument();
        expect(screen.getByText('Add to cart')).toBeInTheDocument();
        expect(screen.getByAltText('imgAlt')).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    it('should call the add to cart with the correct params when the button is clicked', async () => {
        const user = userEvent.setup();

        const onAddToCart = jest.fn();
        const product = {
            id: 'id',
            name: 'name',
            description: 'description',
            price: 10,
            slug: 'slug',
            imgAlt: 'imgAlt',
        };

        render(<CatalogItem product={product} onAddToCart={onAddToCart} />, {
            wrapper: ThemeProvider,
        });

        await userEvent.type(screen.getByLabelText('Quantity'), '{Backspace}2');
        await user.click(screen.getByText('Add to cart'));

        expect(onAddToCart).toHaveBeenCalledWith({
            product,
            quantity: 2,
        });
    });

    it('should not call the callback if the number is out of range', async () => {
        const user = userEvent.setup();

        const onAddToCart = jest.fn();

        render(
            <CatalogItem
                product={{
                    id: 'id',
                    name: 'name',
                    description: 'description',
                    price: 10,
                    slug: 'slug',
                    imgAlt: 'imgAlt',
                }}
                onAddToCart={onAddToCart}
            />,
            {
                wrapper: ThemeProvider,
            },
        );

        await userEvent.type(
            screen.getByLabelText('Quantity'),
            '{Backspace}-2',
        );
        await user.click(screen.getByText('Add to cart'));

        expect(onAddToCart).not.toHaveBeenCalled();
    });

    it('should not call the callback if the value is not a number', async () => {
        const user = userEvent.setup();

        const onAddToCart = jest.fn();

        render(
            <CatalogItem
                product={{
                    id: 'id',
                    name: 'name',
                    description: 'description',
                    price: 10,
                    slug: 'slug',
                    imgAlt: 'imgAlt',
                }}
                onAddToCart={onAddToCart}
            />,
            {
                wrapper: ThemeProvider,
            },
        );

        await userEvent.type(
            screen.getByLabelText('Quantity'),
            'of course this is not a number',
        );
        await user.click(screen.getByText('Add to cart'));

        expect(onAddToCart).not.toHaveBeenCalled();
    });
});
