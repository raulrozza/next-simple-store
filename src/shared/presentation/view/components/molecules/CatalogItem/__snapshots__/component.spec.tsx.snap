// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogItem should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c2 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #754A56;
  color: #FFFFFF;
}

.c2:hover {
  background-color: #845864;
}

.c0 {
  display: grid;
  grid-template-rows: auto 3.5rem 2rem 2rem 2rem;
  background-color: #B3818F;
  overflow: hidden;
  height: 23rem;
  color: #FFFFFF;
  padding-bottom: 0.25rem;
}

.c0 strong {
  font-family: 'Open Sans',sans-serif;
  font-size: 1.5rem;
  margin: 0.25rem;
}

.c0 .price {
  margin: 0rem 0.25rem;
}

.c0 button {
  justify-self: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  margin: 0.25rem;
}

.c1 label {
  font-size: 0.7rem;
}

.c1 input {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  outline: none;
  font-size: 0.7rem;
}

<body>
    <div>
      <li
        class="c0"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
          />
          <img
            alt="imgAlt"
            data-nimg="responsive"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
        <strong>
          name
        </strong>
        <p
          class="price"
        >
          $
          10.00
        </p>
        <div
          class="c1"
        >
          <label
            for="quantity-id"
          >
            Quantity
          </label>
          <input
            id="quantity-id"
            type="number"
            value="1"
          />
        </div>
        <button
          class="c2"
        >
          <svg
            data-testid="button-Icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504.717 320H211.572l6.545 32h268.418c15.401 0 26.816 14.301 23.403 29.319l-5.517 24.276C523.112 414.668 536 433.828 536 456c0 31.202-25.519 56.444-56.824 55.994-29.823-.429-54.35-24.631-55.155-54.447-.44-16.287 6.085-31.049 16.803-41.548H231.176C241.553 426.165 248 440.326 248 456c0 31.813-26.528 57.431-58.67 55.938-28.54-1.325-51.751-24.385-53.251-52.917-1.158-22.034 10.436-41.455 28.051-51.586L93.883 64H24C10.745 64 0 53.255 0 40V24C0 10.745 10.745 0 24 0h102.529c11.401 0 21.228 8.021 23.513 19.19L159.208 64H551.99c15.401 0 26.816 14.301 23.403 29.319l-47.273 208C525.637 312.246 515.923 320 504.717 320zM408 168h-48v-40c0-8.837-7.163-16-16-16h-16c-8.837 0-16 7.163-16 16v40h-48c-8.837 0-16 7.163-16 16v16c0 8.837 7.163 16 16 16h48v40c0 8.837 7.163 16 16 16h16c8.837 0 16-7.163 16-16v-40h48c8.837 0 16-7.163 16-16v-16c0-8.837-7.163-16-16-16z"
            />
          </svg>
          Add to cart
        </button>
      </li>
    </div>
  </body>,
  "container": .c2 {
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  -webkit-transition: background-color 0.2s;
  transition: background-color 0.2s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  background-color: #754A56;
  color: #FFFFFF;
}

.c2:hover {
  background-color: #845864;
}

.c0 {
  display: grid;
  grid-template-rows: auto 3.5rem 2rem 2rem 2rem;
  background-color: #B3818F;
  overflow: hidden;
  height: 23rem;
  color: #FFFFFF;
  padding-bottom: 0.25rem;
}

.c0 strong {
  font-family: 'Open Sans',sans-serif;
  font-size: 1.5rem;
  margin: 0.25rem;
}

.c0 .price {
  margin: 0rem 0.25rem;
}

.c0 button {
  justify-self: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.25rem;
  margin: 0.25rem;
}

.c1 label {
  font-size: 0.7rem;
}

.c1 input {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  outline: none;
  font-size: 0.7rem;
}

<div>
    <li
      class="c0"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 100% 0px 0px 0px;"
        />
        <img
          alt="imgAlt"
          data-nimg="responsive"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <strong>
        name
      </strong>
      <p
        class="price"
      >
        $
        10.00
      </p>
      <div
        class="c1"
      >
        <label
          for="quantity-id"
        >
          Quantity
        </label>
        <input
          id="quantity-id"
          type="number"
          value="1"
        />
      </div>
      <button
        class="c2"
      >
        <svg
          data-testid="button-Icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M504.717 320H211.572l6.545 32h268.418c15.401 0 26.816 14.301 23.403 29.319l-5.517 24.276C523.112 414.668 536 433.828 536 456c0 31.202-25.519 56.444-56.824 55.994-29.823-.429-54.35-24.631-55.155-54.447-.44-16.287 6.085-31.049 16.803-41.548H231.176C241.553 426.165 248 440.326 248 456c0 31.813-26.528 57.431-58.67 55.938-28.54-1.325-51.751-24.385-53.251-52.917-1.158-22.034 10.436-41.455 28.051-51.586L93.883 64H24C10.745 64 0 53.255 0 40V24C0 10.745 10.745 0 24 0h102.529c11.401 0 21.228 8.021 23.513 19.19L159.208 64H551.99c15.401 0 26.816 14.301 23.403 29.319l-47.273 208C525.637 312.246 515.923 320 504.717 320zM408 168h-48v-40c0-8.837-7.163-16-16-16h-16c-8.837 0-16 7.163-16 16v40h-48c-8.837 0-16 7.163-16 16v16c0 8.837 7.163 16 16 16h48v40c0 8.837 7.163 16 16 16h16c8.837 0 16-7.163 16-16v-40h48c8.837 0 16-7.163 16-16v-16c0-8.837-7.163-16-16-16z"
          />
        </svg>
        Add to cart
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
